README.md
*©2023 Axis Communications AB. AXIS COMMUNICATIONS, AXIS, ARTPEC and VAPIX are registered trademarks of Axis AB in various jurisdictions. All other trademarks are the property of their respective owners.*

<!-- omit in toc -->
# Application for Amazon Kinesis Video Streams

The Application for Amazon Kinesis Video Streams can be run on a virtualization enabled
Axis camera as a container, making it possible to stream video to
[Amazon Kinesis Video Streams](https://aws.amazon.com/kinesis/video-streams/). The
stream can thereafter be fed into other AWS services such as [Amazon Rekognition](https://aws.amazon.com/rekognition/) to
perform image and or video analytics.

![diagram](./assets/diagram.png)

<!-- omit in toc -->
## Table of contents

- [Target Audience](#target-audience)
- [Requirements](#requirements)
- [Option 1: Access Kinesis Video Streams resources using IAM](#option-1-access-kinesis-video-streams-resources-using-iam)
    - [Variables](#variables)
        - [Container Build Variables](#container-build-variables)
        - [Container Runtime Variables](#container-runtime-variables)
    - [Install](#install)
        - [From Docker Hub](#from-docker-hub)
        - [Build Locally](#build-locally)
- [Option 2: Access Kinesis Video Streams resources using AWS IoT](#option-2-access-kinesis-video-streams-resources-using-aws-iot)
    - [Prerequisites](#prerequisites)
    - [Creating the Certificate Files](#creating-the-certificate-files)
    - [Install](#install-1)
        - [From Docker Hub](#from-docker-hub-1)
        - [Build Locally](#build-locally-1)
- [Run on the Camera](#run-on-the-camera)
    - [Save and Load the Image to the Camera](#save-and-load-the-image-to-the-camera)
    - [Starting the Container](#starting-the-container)
- [Use case - Event based Amazon Kinesis Video Streams](#use-case---event-based-amazon-kinesis-video-streams)
- [Verify That the Amazon Kinesis Video Stream is Successfully Running](#verify-that-the-amazon-kinesis-video-stream-is-successfully-running)
- [Known Limitations](#known-limitations)
- [Disclaimer](#disclaimer)
- [License](#license)

## Target Audience

This repository is intended for [low code developers](https://en.wikipedia.org/wiki/Low-code_development_platform) building integration
offerings to connect and stream Axis cameras video into the Amazon Kinesis Video Streams service.

## Requirements

The following setup is supported:

- Camera
    - Chip: ARTPEC-{7-8} [DLPU enabled devices](https://axiscommunications.github.io/acap-documentation/docs/axis-devices-and-compatibility/#acap-computer-vision-sdk) (e.g., Q1615 MkIII, P3265-LV)
    - Firmware: 10.9 or higher
    - [Docker ACAP](https://github.com/AxisCommunications/docker-acap) installed and started, using TLS and SD card as storage

- Computer
    - OS: Linux/macOS running preferred shell, or Windows with WSL2 installed to run Bash on Windows
    - AWS Account with [security credentials](https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_access-keys.html) generated
        - [Option 1: Access Kinesis Video Streams resources using IAM](#option-1-access-kinesis-video-streams-resources-using-iam)
            - Access key ID
            - Secret access key
        - [Option 2: Access Kinesis Video Streams resources using AWS IoT](#option-2-access-kinesis-video-streams-resources-using-aws-iot)
            - Option 2.1. Using AWS IoT certificate alone
            - Option 2.2. Generating temporary credentials (temporary access key ID, secret access key and session token) from the IoT certificate
    - AWS CLI
        - [Getting started with the AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/cli-chap-getting-started.html)
            - Ensure to choose a region that supports Amazon Kinesis Video Streams.
    - [Docker](https://docs.docker.com/get-docker/) with BuildKit enabled
    - [Docker Compose](https://docs.docker.com/compose/install/)
    - Only required for option 2: [jq](https://stedolan.github.io/jq/), a lightweight command-line JSON processor

> **Note** Developers are recommended to follow [AWS Credentials access best practices](https://docs.aws.amazon.com/accounts/latest/reference/credentials-access-keys-best-practices.html) in safeguarding the credentials included
> in developer machine configuration.

## Option 1: Access Kinesis Video Streams resources using IAM

When setting up an IAM user or a role for streaming to Amazon Kinesis Video Streams, see the [official AWS documentation](https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/how-iam.html) describing how to define permissions.

### Variables

To run the solution, a number of variables need to be added. These will be used for building or pulling the Docker image and
running it.

#### Container Build Variables

Add the IMAGE_NAME as a shell variable so that it can be reused:

```sh
export IMAGE_NAME=axisecp/kinesis-video-stream-application
```

Also, add the IMAGE_TAG:

```sh
export IMAGE_TAG=latest-<armv7hf or aarch64>
```

where the IMAGE_TAG is `latest-armv7hf` for ARTPEC-7 and `latest-aarch64` for
ARTPEC-8 devices.

#### Container Runtime Variables

Before running the solution, environment variables need to be set up.
Create a file named `.env` in the root directory of this repository, it will contain data to communicate with the camera and
AWS. After creating the file, add the content below to the file and fill in the corresponding values:

```sh
# Camera specific variables
DEVICE_USERNAME=<camera username>
DEVICE_PASSWORD=<camera password>

# AWS related variables
AWS_KINESIS_STREAM_NAME=<Amazon Kinesis video stream name>
AWS_REGION=<AWS region>
AWS_ACCESS_KEY_ID=<AWS access key ID>
AWS_SECRET_ACCESS_KEY=<AWS secret key>
```

### Install

The image can be retrieved by either pulling it from Docker Hub, or by building
it locally.

#### From Docker Hub

Get the Docker image by pulling it from Docker Hub:

```sh
docker pull ${IMAGE_NAME}:${IMAGE_TAG}
```

#### Build Locally

Add the architecture for the Docker image as a shell variable, depending on the camera
system-on-chip. Use `arm32v7` for ARTPEC-7 devices:

```sh
export ARCH=arm32v7
```

and `arm64v8` for ARTPEC-8:

```sh
export ARCH=arm64v8
```

Once the shell variables have been added, the Docker image can be built:

```sh
docker buildx build --tag ${IMAGE_NAME}:${IMAGE_TAG} --build-arg ARCH --build-arg KVS_CPP_PRODUCER_SDK_TAG=v3.3.1 .
```

## Option 2: Access Kinesis Video Streams resources using AWS IoT

Amazon Kinesis Video Streams do not support certificate-based authentication, however, AWS IoT has a credentials provider that allows
you to use the built-in X.509 certificate as the unique device identity to authenticate AWS requests.

### Prerequisites

To create the certificate, the following AWS resources need to be set up.

- An IoT Thing Type and an IoT Thing.
- An IAM Role to be assumed by IoT.
- X.509 certificate created and configured.

AWS provides [documentation](https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/how-iot.html) for how to set the above
prerequisites up. However, to simplify the process of setting up an IAM Role assumed by IoT, policies, certificate etc, a [script](https://github.com/AxisCommunications/axis-aws-kinesis-vsp/blob/main/x509-authentication/certificate_files/generate_certificate.sh)
is provided.

If the prerequisites are set up by following the
[documentation](https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/how-iot.html) from AWS the step [Creating the certificate files](#creating-the-certificate-files) below can be skipped.

### Creating the Certificate Files

1. Create a file named `.env` in the `x509-authentication` directory of this repository, it will contain data to communicate with
    the camera and AWS. Add the content below to the file and fill in the corresponding values:

    > **Note** For the `generate_certificate.sh` script and Docker Compose to pick up the environment variables the file needs to be named `.env`
    **and** be placed in the `x509-authentication` directory.

    ```sh
    DEVICE_USERNAME=<camera username>
    DEVICE_PASSWORD=<camera password>
    GST_PLUGIN_PATH=/opt/app/amazon-kinesis-video-streams-producer-sdk-cpp/build
    # PROXY=<proxy address, needed if camera is behind a proxy>

    # AWS related variables
    AWS_KINESIS_STREAM_NAME=<Amazon Kinesis video stream name>
    AWS_REGION=<AWS region>
    AWS_CREDENTIAL_PROVIDER=<AWS credential provider, see next step for how to fetch it>
    AWS_ROLE_ALIAS=<Pick a name for the AWS Role Alias>

    # Additional AWS variables needed for generating certificate
    AWS_THING=$AWS_KINESIS_STREAM_NAME
    AWS_THING_TYPE=<Pick a name for the AWS Thing Type>
    AWS_ROLE=<Pick a name for the AWS Role>
    AWS_IAM_POLICY=<Pick a name for the AWS IAM Policy>
    AWS_IOT_POLICY=<Pick a name for the AWS IOT Policy>
    AWS_ROOT_CA_ADDRESS=https://www.amazontrust.com/repository/SFSRootCAG2.pem
    ```

    It's required that the value of `AWS_THING` is identical to the value of `AWS_KINESIS_STREAM_NAME`.

    To fetch the credentials provider the following command can be used:

    ```sh
    aws --profile default iot describe-endpoint --endpoint-type iot:CredentialProvider --output text
    ```

2. Step into the `certificate_files` directory and run the following command to generate certificate and keys:

    > **Note** This step can be skipped if setting up the certificate manually according to the [AWS documentation](https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/how-iot.html).

    If `AWS CLI` is configured with `output = json` the script can be run as:

    ```sh
    ./generate_certificate.sh ../.env
    ```

    If another configuration is set, `AWS_DEFAULT_OUTPUT="json"` can be added to the call instead, i.e.:

    ```sh
    AWS_DEFAULT_OUTPUT="json" ./generate_certificate.sh ../.env
    ```

    > **Note** Currently the script only supports a profile named `default` for the calls made towards AWS CLI. If you like it to use another profile read about [Named profiles for the AWS CLI](https://docs.aws.amazon.com/cli/latest/userguide/cli-configure-profiles.html) in the AWS documentation and update the script accordingly.

### Install

There's two options for how to run an image using the generated certificate. Either by building the certificate into an image or by generating temporary credentials from the certificate.

- To be able to build or pull the image, set the following environment variables in your shell:

    ```sh
    export IMAGE_NAME=axisecp/kinesis-video-stream-application
    export IMAGE_TAG=<Choose either latest-armv7hf or latest-aarch64>
    export DEVICE_IP=<camera IP>
    ```

#### From Docker Hub

With this option the image need **not** be rebuilt. To use this option the following steps needs to be performed:

1. Generate the temporary credentials by running the `generate_temporary_credentials.sh` script in the `x509-authentication/certificate` folder similar to how the certificate was generated:

    ```sh
    ./generate_temporary_credentials.sh ../.env
    ```

    The script will echo the temporary credentials and environment variables to set:

    ```sh
    ./generate_temporary_credentials.sh ../.env
    Temporary credentials created with an expiration date set to:2023-02-15T19:54:53Z

    Run the following command in your shell to export the temporary credentials so that they can be picked up by Docker Compose.
    export AWS_ACCESS_KEY_ID=********************
    export AWS_SECRET_ACCESS_KEY=******************
    export AWS_SESSION_TOKEN=************************
    ```

2. Follow the instructions and export the variables:

    ```sh
    export AWS_ACCESS_KEY_ID=********************
    export AWS_SECRET_ACCESS_KEY=******************
    export AWS_SESSION_TOKEN=************************
    ```

    > **Note** The actual values are much longer than the substitutes above `*****`

3. When using temporary credentials, the same image as the one from [Option 1: Access Kinesis Video Streams resources using IAM](#option-1-access-kinesis-video-streams-resources-using-iam) can be used.
Pull the image from DockerHub by running the following command:

    ```sh
    docker pull ${IMAGE_NAME}:${IMAGE_TAG}
    ```

4. Copy the recently created .env file and paste it into the root directory of the repository, so that Docker Compose can pick up all the required variables.

5. Change directory to the root folder of the repository. This is needed to run the application from the correct directory in the 'Run on the Camera' section below.

#### Build Locally

With this option the image **need to be rebuilt** to include the certificate. To use this option the following steps needs to be performed:

1. Step back into the `x509-authentication` directory and build an image:

    ```sh
    docker buildx build --tag ${IMAGE_NAME}:${IMAGE_TAG} --build-arg IMAGE_TAG=$IMAGE_TAG .
    ```

2. Create the Amazon Kinesis Video Stream.

    If the policy is set with the script above or according to the [AWS documentation](https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/how-iot.html),
    permission will **not** be set for `KinesisVideo:CreateStream` action. I.e. the stream will have to be created manually.

    ```sh
    aws kinesisvideo create-stream --data-retention-in-hours 2 --stream-name <name of the stream used in above steps>
    ```

    > **Note** The stream name must be the same as the name of the Thing created earlier.

## Run on the Camera

### Save and Load the Image to the Camera

Add the camera's IP address as a shell variable:

```sh
export DEVICE_IP=<camera IP>
```

Clear Docker memory:

```sh
docker --tlsverify -H $DEVICE_IP:2376 system prune --all --force
```

If you encounter any TLS related issues, please see the TLS setup chapter regarding the `DOCKER_CERT_PATH` environment variable
in the [Docker ACAP repository](https://github.com/AxisCommunications/docker-acap).

The image can now be saved and loaded to the camera:

```sh
docker save ${IMAGE_NAME}:${IMAGE_TAG} | docker --tlsverify -H $DEVICE_IP:2376 load
```

### Starting the Container

To start the container you can use Docker Compose:

```sh
docker --tlsverify -H $DEVICE_IP:2376 compose up
```

or:

```sh
docker --tlsverify -H $DEVICE_IP:2376 compose up -d
```

to run in detached (background) mode.

Once the `docker compose` command has been run, an RTSP stream is set up with the
`start_stream.sh` script. The AWS GStreamer plugin kvssink sends media based on
the Matroska container format to Amazon Kinesis Video Streams. The plugin is set up
with default values, however these values can be modified according to the
[kvssink parameter reference](https://docs.aws.amazon.com/kinesisvideostreams/latest/dg/examples-gstreamer-plugin-parameters.html)
.

## Use case - Event based Amazon Kinesis Video Streams

This repository also includes an example use case [start-on-app-event](./start-on-app-event) that generates [Amazon Kinesis Video Streams](https://aws.amazon.com/kinesis/video-streams/) based on an event triggered by rules created with Axis analytics applications like [AXIS Object Analytics](https://www.axis.com/products/axis-object-analytics) or
[AXIS Video Motion Detection](https://www.axis.com/products/axis-video-motion-detection).

## Verify That the Amazon Kinesis Video Stream is Successfully Running

The most straightforward way to verify that the stream from the camera actually
reaches Amazon Kinesis Video Streams is to do it from the AWS UI.

1. Log in to your AWS account.
2. Search for and go to the Amazon Kinesis Video Streams service.
3. Make sure that you are in the correct AWS region, and select the Amazon Kinesis
video stream in the list.
4. Click the **Media Playback** button.
5. If everything is set up correctly, the stream should show up. Wait a number
of seconds since there might be a delay.

## Known Limitations

When streaming to Amazon Kinesis Video Streams there is a latency which can be
affected by the selected AWS region, network setup and video resolution.

## Disclaimer

<!-- textlint-disable -->

This document and its content are provided courtesy of Axis and all rights to the document shall remain vested in Axis Communications AB. AXIS COMMUNICATIONS, AXIS, ARTPEC and VAPIX are registered trademarks of Axis AB in various jurisdictions. Amazon Web Services, AWS and the Powered by AWS logo are trademarks of Amazon.com, Inc. or its affiliates. Docker and the Docker logo are trademarks or registered trademarks of Docker, Inc. in the United States and/or other countries. Docker, Inc. and other parties may also have trademark rights in other terms used herein. All other trademarks are the property of their respective owners, and we are not affiliated with, endorsed or sponsored by them or their affiliates.

As described in this document, you may be able to connect to, access and use third party products, web sites, example code, software or services (“Third Party Services”). You acknowledge that any such connection and access to such Third Party Services are made available to you for convenience only. Axis does not endorse any Third Party Services, nor does Axis make any representations or provide any warranties whatsoever with respect to any Third Party Services, and Axis specifically disclaims any liability or obligations with regard to Third Party Services. The Third Party Services are provided to you in accordance with their respective terms and conditions, and you alone are responsible for ensuring that you (a) procure appropriate rights to access and use any such Third Party Services and (b) comply with the terms and conditions applicable to its use.

PLEASE BE ADVISED THAT THIS DOCUMENT IS PROVIDED “AS IS” WITHOUT WARRANTY OF ANY KIND, AND IS NOT INTENDED TO, AND SHALL NOT, CREATE ANY LEGAL OBLIGATION FOR AXIS COMMUNICATIONS AB AND/OR ANY OF ITS AFFILIATES. THE ENTIRE RISK AS TO THE USE, RESULTS AND PERFORMANCE OF THIS DOCUMENT AND ANY THIRD PARTY SERVICES REFERENCED HEREIN IS ASSUMED BY THE USER OF THE DOCUMENT AND AXIS DISCLAIMS AND EXCLUDES, TO THE MAXIMUM EXTENT PERMITTED BY LAW, ALL WARRANTIES, WHETHER STATUTORY, EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO ANY IMPLIED WARRANTIES OF MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, TITLE AND NON-INFRINGEMENT AND PRODUCT LIABILITY.

<!-- textlint-enable -->

## License

Example Code means the examples provided by Axis in this document within the grey text boxes.

Example Code ©2023 Axis Communications AB is licensed under the [Apache License, Version 2.0 (the “License”)](./LICENSE). You may not use the Example Code except in compliance with the License.

You may obtain a copy of the License at <https://www.apache.org/licenses/LICENSE-2.0>. Example Code distributed under the License is distributed on an “AS IS” BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

This is a list of the repository's releases and the changes included in them. Each release version follows the semantic
versioning ([SemVer](https://semver.org/)) scheme.

## [unreleased]

## [v1.0.0]

### Added

- First Release for "Application for Amazon Kinesis Video Streams" including an example use case of streaming Amazon Kinesis video based on events triggered by Axis analytics applications.


get_kvs_clip.sh
#!/bin/bash

# Parse the input arguments
while [[ "$#" -gt 0 ]]; do
    case $1 in
        --clip-length) clip_length="$2"; shift ;;
        --use-tls) use_tls="$2"; shift ;;
    esac
    shift
done

# Verify mandatory parameters
if [ -z "$clip_length" ]; then
    echo "Input argument --clip-length is missing, please add it:"
    echo "--clip-length <lenght of clip in seconds to stream and download>"
    exit 1
fi

# Export env vars
set -o allexport
# shellcheck source=/dev/null
source .env
set +o allexport

echo "Getting temporary credentials"
aws_credential_provider=$(aws --profile default iot describe-endpoint --endpoint-type iot:CredentialProvider --output text)
token=$(curl --silent -H "x-amzn-iot-thingname:${AWS_KINESIS_STREAM_NAME}" --cert ${CERTIFICATE_PATH}/certificate.pem --key ${CERTIFICATE_PATH}/private.pem.key https://${aws_credential_provider}/role-aliases/${AWS_ROLE_ALIAS}/credentials --cacert ${CERTIFICATE_PATH}/cacert.pem)

# Export additional varaiables needed to run the commands towards AWS
export AWS_ACCESS_KEY_ID_TMP=$(jq --raw-output '.credentials.accessKeyId' <<< $token)
export AWS_SECRET_ACCESS_KEY_TMP=$(jq --raw-output '.credentials.secretAccessKey' <<< $token)
export AWS_SESSION_TOKEN_TMP=$(jq --raw-output '.credentials.sessionToken' <<< $token)

# Start the stream
echo "Starting Kinesis Video Stream"
if [ -z "$use_tls" ]; then
    docker-compose -H $DEVICE_IP:2375 up -d
else
    docker-compose --tlsverify -H $DEVICE_IP:2376 up -d
fi

# Get data endpoint for downloading clip
data_endpoint_json=$(aws kinesisvideo get-data-endpoint --stream-name ${AWS_KINESIS_STREAM_NAME} --api-name GET_DASH_STREAMING_SESSION_URL --region ${AWS_REGION})
data_endpoint=$(jq --raw-output '.DataEndpoint' <<< $data_endpoint_json)

# Wait for stream to finish
echo "Running stream for ${clip_length} seconds..."
sleep $((clip_length + 10))
fragment_selector_json='{"FragmentSelectorType": "SERVER_TIMESTAMP", "TimestampRange": {"StartTimestamp": '$(expr $(date +%s) - ${clip_length})', "EndTimestamp": '$(date +%s)'}}'

# Download clip
echo "Downloading video clip..."
aws kinesis-video-archived-media get-clip --stream-name ${AWS_KINESIS_STREAM_NAME} --clip-fragment-selector "${fragment_selector_json}" --endpoint-url ${data_endpoint} ./video.mp4

# Stop the stream
echo "Stopping stream"
if [ -z "$use_tls" ]; then
    docker-compose -H $DEVICE_IP:2375 down
else
    docker-compose --tlsverify -H $DEVICE_IP:2376 down
fi


start-on-app-event/app.py
""""
    Copyright (C) 2023 Axis Communications AB, Lund, Sweden

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

        <http://www.apache.org/licenses/LICENSE-2.0>

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
"""

import json
import os
import time
import xml.etree.ElementTree as ET
from ast import literal_eval
from datetime import date

import requests
from requests.auth import HTTPDigestAuth

app_name = os.environ["APPNAME"]
device_ip = os.environ["DEVICE_IP"]
device_username = os.environ["DEVICE_USERNAME"]
device_password = os.environ["DEVICE_PASSWORD"]
generate_event_triggers = literal_eval(
    os.environ["GENERATE_EVENT_TRIGGERS"].lower().capitalize()
)


def get_recordings_path():
    date_today = str(date.today().isoformat()).replace("-", "")
    return "/tmp/recordings/" + date_today


def get_file_data(file_name):
    f = open(file_name)
    file_data = f.read()
    f.close()
    return file_data


def set_configuration(data):
    url = "http://{}/local/{}/control.cgi".format(device_ip, app_name)
    requests.post(url, auth=HTTPDigestAuth(device_username, device_password), json=data)


def vapix_services(xml):
    url = "http://{}/vapix/services".format(device_ip)
    response = requests.post(
        url,
        auth=HTTPDigestAuth(device_username, device_password),
        headers={"Content-Type": "text/plain;charset=UTF-8"},
        data=xml.encode("utf-8"),
    )
    return response.text


def get_action_configuration_id(response):
    root = ET.fromstring(response)
    for element in root.iter():
        if "ConfigurationID" in element.tag:
            return element.text


def get_folder_composition(path):
    file_path_set = set()
    for dir_path, _, files in os.walk(path):
        file_paths = list(map(lambda f: os.path.join(dir_path, f), files))
        mkv_file_paths = filter(lambda f_p: f_p.endswith(".mkv"), file_paths)
        file_path_set = file_path_set.union(set(mkv_file_paths))
    return file_path_set


def folder_has_changed(current, last):
    return len(current - last) > 0


def generate_rules_and_actions():
    app_config_data = get_file_data("json/" + app_name + "_config.json")
    app_config_data_json = json.loads(app_config_data)
    set_configuration(app_config_data_json)

    add_action_configuration = get_file_data("xml/add_action_configuration.xml")
    time.sleep(5)
    add_action_rule = get_file_data("xml/" + app_name + "_add_action_rule.xml")
    response = vapix_services(add_action_configuration)

    action_config_id = get_action_configuration_id(response)
    new_action_rule = add_action_rule.replace("ConfigurationID", str(action_config_id))
    vapix_services(new_action_rule)


if generate_event_triggers:
    generate_rules_and_actions()

recordings_path = get_recordings_path()
last_recordings_path = recordings_path

last_folder_comp = get_folder_composition(recordings_path)

while True:
    recordings_path = get_recordings_path()
    current_folder_comp = get_folder_composition(recordings_path)

    if recordings_path != last_recordings_path:
        last_recordings_path = recordings_path
        last_folder_comp = current_folder_comp
    else:
        if folder_has_changed(current_folder_comp, last_folder_comp):
            new_file_path = str(sorted(list(current_folder_comp))[-1])
            os.environ["FILE_NAME"] = new_file_path
            os.system("./start_stream_mkv.sh")
        else:
            print("no event detected...")
        last_folder_comp = current_folder_comp
        last_recordings_path = recordings_path
        time.sleep(3)


X509_authentication/certificate_files/generate.sh
#!/bin/bash

ENV_FILE_PATH="$1"

# Export env vars
set -o allexport
# shellcheck source=/dev/null
source "$ENV_FILE_PATH"
set +o allexport

# Create a temporary directory for generated files
mkdir tmp

echo "Creating IoT Thing: ${AWS_THING} and Thing Type: ${AWS_THING_TYPE}"
aws --profile default iot create-thing-type --thing-type-name "${AWS_THING_TYPE}" > tmp/iot-thing-type.json
aws --profile default iot create-thing --thing-name "${AWS_THING}" --thing-type-name "${AWS_THING_TYPE}" > tmp/iot-thing.json


cat > tmp/iam-policy-document.json <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "Service": "credentials.iot.amazonaws.com"
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
EOF

echo ""
echo "Creating IAM Role: ${AWS_ROLE}"
if aws --profile default iam get-role --role-name "${AWS_ROLE}" > tmp/iam-role.json 2>&1; then
    echo "Role: ${AWS_ROLE} already available, continuing..."
else
    aws --profile default iam create-role --role-name "${AWS_ROLE}" --assume-role-policy-document 'file://tmp/iam-policy-document.json' > tmp/iam-role.json
fi


cat > tmp/iam-permission-document.json <<EOF
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "kinesisvideo:DescribeStream",
                "kinesisvideo:PutMedia",
                "kinesisvideo:TagStream",
                "kinesisvideo:GetDataEndpoint"
            ],
            "Resource": "arn:aws:kinesisvideo:*:*:stream/\${credentials-iot:ThingName}/*"
        }
    ]
}
EOF

echo "Attaching permissions policy: ${AWS_IAM_POLICY} to: ${AWS_ROLE}"
aws --profile default iam put-role-policy --role-name "${AWS_ROLE}" --policy-name "${AWS_IAM_POLICY}" --policy-document 'file://tmp/iam-permission-document.json'

# Create a Role Alias
echo ""
echo "Creating Role Alias: ${AWS_ROLE_ALIAS}"
if aws --profile default iot describe-role-alias --role-alias "${AWS_ROLE_ALIAS}" > tmp/iot-role-alias.json 2>&1; then
    echo "Role Alias: ${AWS_ROLE_ALIAS} already available, continuing..."
else
    aws --profile default iot create-role-alias --role-alias "${AWS_ROLE_ALIAS}" --role-arn "$(jq --raw-output '.Role.Arn' tmp/iam-role.json)" --credential-duration-seconds 3600
    aws --profile default iot describe-role-alias --role-alias "${AWS_ROLE_ALIAS}" > tmp/iot-role-alias.json
fi


cat > tmp/iot-policy-document.json <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "iot:Connect"
      ],
      "Resource": "$(jq --raw-output '.roleAliasDescription.roleAliasArn' tmp/iot-role-alias.json)"
    },
    {
      "Effect": "Allow",
      "Action": [
        "iot:AssumeRoleWithCertificate"
      ],
      "Resource": "$(jq --raw-output '.roleAliasDescription.roleAliasArn' tmp/iot-role-alias.json)"
    }
  ]
}
EOF

echo ""
echo "Creating Policy: ${AWS_IOT_POLICY} that will enable IoT to assume role with the certificate (once it is attached) using the role alias"
if aws --profile default iot get-policy --policy-name "${AWS_IOT_POLICY}" 2>/dev/null; then
    echo "Policy: ${AWS_IOT_POLICY} already available, continuing..."
else
    aws --profile default iot create-policy --policy-name "${AWS_IOT_POLICY}" --policy-document 'file://tmp/iot-policy-document.json'
fi

echo ""
echo "Creating the certificate to which the policy for IoT created above must be attached"
aws --profile default iot create-keys-and-certificate --set-as-active --certificate-pem-outfile certificate.pem --public-key-outfile public.pem.key --private-key-outfile private.pem.key > certificate

echo "Attaching the policy for IoT (${AWS_IOT_POLICY} created in earlier steps) to the certificate"
aws --profile default iot attach-policy --policy-name "${AWS_IOT_POLICY}" --target "$(jq --raw-output '.certificateArn' certificate)"

echo "Attaching the IoT thing ${AWS_THING} to the certificate"
aws --profile default iot attach-thing-principal --thing-name "${AWS_THING}" --principal "$(jq --raw-output '.certificateArn' certificate)"

echo ""
echo "Getting endpoint needed to authorize requests through the IoT credentials provider"
aws --profile default iot describe-endpoint --endpoint-type iot:CredentialProvider --output text > tmp/iot-credential-provider.txt

echo "Fetching CA certificate needed to establish trust with the back-end service"
curl --silent "${AWS_ROOT_CA_ADDRESS}" --output cacert.pem

# Clean up tmp folder
rm -rf tmp


x509-authentication/certificate_files/generate_certificate.sh
#!/bin/bash

if [ -z "$1" ]; then
        echo "Please add the path to the .env file as an argument"
        exit 1
fi

ENV_FILE_PATH="$1"

# Export env vars
set -o allexport
# shellcheck source=/dev/null
source "$ENV_FILE_PATH"
set +o allexport

# Create a temporary directory for generated files
mkdir tmp

echo "Creating IoT Thing: ${AWS_THING} and Thing Type: ${AWS_THING_TYPE}"
aws --profile default iot create-thing-type --thing-type-name "${AWS_THING_TYPE}" >tmp/iot-thing-type.json
aws --profile default iot create-thing --thing-name "${AWS_THING}" --thing-type-name "${AWS_THING_TYPE}" >tmp/iot-thing.json

cat >tmp/iam-policy-document.json <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Principal": {
        "Service": "credentials.iot.amazonaws.com"
      },
      "Action": "sts:AssumeRole"
    }
  ]
}
EOF

echo ""
echo "Creating IAM Role: ${AWS_ROLE}"
if aws --profile default iam get-role --role-name "${AWS_ROLE}" >tmp/iam-role.json 2>&1; then
        echo "Role: ${AWS_ROLE} already available, continuing..."
else
        aws --profile default iam create-role --role-name "${AWS_ROLE}" --assume-role-policy-document 'file://tmp/iam-policy-document.json' >tmp/iam-role.json
fi

cat >tmp/iam-permission-document.json <<EOF
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "kinesisvideo:DescribeStream",
                "kinesisvideo:PutMedia",
                "kinesisvideo:TagStream",
                "kinesisvideo:GetDataEndpoint"
            ],
            "Resource": "arn:aws:kinesisvideo:*:*:stream/\${credentials-iot:ThingName}/*"
        }
    ]
}
EOF

echo "Attaching permissions policy: ${AWS_IAM_POLICY} to: ${AWS_ROLE}"
aws --profile default iam put-role-policy --role-name "${AWS_ROLE}" --policy-name "${AWS_IAM_POLICY}" --policy-document 'file://tmp/iam-permission-document.json'

# Create a Role Alias
echo ""
echo "Creating Role Alias: ${AWS_ROLE_ALIAS}"
if aws --profile default iot describe-role-alias --role-alias "${AWS_ROLE_ALIAS}" >tmp/iot-role-alias.json 2>&1; then
        echo "Role Alias: ${AWS_ROLE_ALIAS} already available, continuing..."
else
        aws --profile default iot create-role-alias --role-alias "${AWS_ROLE_ALIAS}" --role-arn "$(jq --raw-output '.Role.Arn' tmp/iam-role.json)" --credential-duration-seconds 3600
        aws --profile default iot describe-role-alias --role-alias "${AWS_ROLE_ALIAS}" >tmp/iot-role-alias.json
fi

cat >tmp/iot-policy-document.json <<EOF
{
  "Version": "2012-10-17",
  "Statement": [
    {
      "Effect": "Allow",
      "Action": [
        "iot:Connect"
      ],
      "Resource": "$(jq --raw-output '.roleAliasDescription.roleAliasArn' tmp/iot-role-alias.json)"
    },
    {
      "Effect": "Allow",
      "Action": [
        "iot:AssumeRoleWithCertificate"
      ],
      "Resource": "$(jq --raw-output '.roleAliasDescription.roleAliasArn' tmp/iot-role-alias.json)"
    }
  ]
}
EOF

echo ""
echo "Creating Policy: ${AWS_IOT_POLICY} that will enable IoT to assume role with the certificate (once it is attached) using the role alias"
if aws --profile default iot get-policy --policy-name "${AWS_IOT_POLICY}" 2>/dev/null; then
        echo "Policy: ${AWS_IOT_POLICY} already available, continuing..."
else
        aws --profile default iot create-policy --policy-name "${AWS_IOT_POLICY}" --policy-document 'file://tmp/iot-policy-document.json'
fi

echo ""
echo "Creating the certificate to which the policy for IoT created above must be attached"
aws --profile default iot create-keys-and-certificate --set-as-active --certificate-pem-outfile certificate.pem --public-key-outfile public.pem.key --private-key-outfile private.pem.key >certificate

echo "Attaching the policy for IoT (${AWS_IOT_POLICY} created in earlier steps) to the certificate"
aws --profile default iot attach-policy --policy-name "${AWS_IOT_POLICY}" --target "$(jq --raw-output '.certificateArn' certificate)"

echo "Attaching the IoT thing ${AWS_THING} to the certificate"
aws --profile default iot attach-thing-principal --thing-name "${AWS_THING}" --principal "$(jq --raw-output '.certificateArn' certificate)"

echo ""
echo "Getting endpoint needed to authorize requests through the IoT credentials provider"
aws --profile default iot describe-endpoint --endpoint-type iot:CredentialProvider --output text >tmp/iot-credential-provider.txt

echo "Fetching CA certificate needed to establish trust with the back-end service"
curl --silent "${AWS_ROOT_CA_ADDRESS}" --output cacert.pem

# Clean up tmp folder
rm -rf tmp
